src/storage/store/var_list_column.cpp

```diff
@@ -175,29 +275,110 @@ offset_t VarListColumn::readOffset(
+            auto localUpdateChunk = localUpdateChunks[chunkIdx];
+            dstOffsets.push_back(offsetInDstChunk);
+            columnChunk->append(localUpdateChunk, offsetInLocalChunk, 1);
+        }
+        for (auto& [offsetInDstChunk, rowIdx] : insertInfo) {
+            auto [chunkIdx, offsetInLocalChunk] =
+                LocalChunkedGroupCollection::getChunkIdxAndOffsetInChunk(rowIdx);
+            auto localInsertChunk = localInsertChunks[chunkIdx];
+            dstOffsets.push_back(offsetInDstChunk);
+            columnChunk->append(localInsertChunk, offsetInLocalChunk, 1);
+        }
+        prepareCommitForChunk(transaction, nodeGroupIdx, dstOffsets, columnChunk.get(), 0);
+    }
+}
+
+void VarListColumn::prepareCommitForChunk(Transaction* transaction, node_group_idx_t nodeGroupIdx,
+    const std::vector<common::offset_t>& dstOffsets, ColumnChunk* chunk, offset_t startSrcOffset) {
+    auto currentNumNodeGroups = metadataDA->getNumElements(transaction->getType());
+    auto isNewNodeGroup = nodeGroupIdx >= currentNumNodeGroups;
+    if (isNewNodeGroup) {
+        commitColumnChunkOutOfPlace(
+            transaction, nodeGroupIdx, isNewNodeGroup, dstOffsets, chunk, startSrcOffset);
+    } else {
+        // we separate the commit into three parts: offset chunk commit, size column chunk commit,
+        // data column chunk
+        auto varListChunk = ku_dynamic_cast<ColumnChunk*, VarListColumnChunk*>(chunk);
+        sizeColumn->prepareCommitForChunk(transaction, nodeGroupIdx, dstOffsets,
+            varListChunk->getSizeColumnChunk(), startSrcOffset);
+        auto dataColumnSize =
+            dataColumn->getMetadata(nodeGroupIdx, transaction->getType()).numValues;
+        auto dataColumnChunk = varListChunk->getDataColumnChunk();
+        auto appendListNum = std::min(chunk->getNumValues(), (uint64_t)dstOffsets.size());
+        auto appendListDataNum = 0u;
+        auto startVarListOffset = varListChunk->getListStartOffset(startSrcOffset);
+        std::vector<common::offset_t> dstOffsetsInDataColumn;
+        for (auto i = 0u; i < appendListNum; i++) {
+            auto appendLen = varListChunk->getListLen(startSrcOffset + i);
+            for (auto j = 0u; j < appendLen; j++) {
+                dstOffsetsInDataColumn.push_back(dataColumnSize + appendListDataNum + j);
+            }
+            appendListDataNum += appendLen;
+        }
+        dataColumn->prepareCommitForChunk(
+            transaction, nodeGroupIdx, dstOffsetsInDataColumn, dataColumnChunk, startVarListOffset);
+        // we need to update the offset since we do not do in-place list data update but append data
+        // in the end of list data column we need to plus to original data column size to get the
+        // new offset
+        // TODO(Jiamin): A better way is to store the offset in a offset column, just like size
+        // column. Then we can reuse prepareCommitForChunk interface for offset column.
```

Comment:
We should apply this optimization. You can separate it in another PR, but we should do it soon.